core/src/main/java/io/questdb/griffin/engine/groupby/DirectStrAwareCharHolder.java

```diff
@@ -0,0 +1,166 @@
+ * direct, then the Holder stores only a pointer to the char sequence and its length. Otherwise, it stores the char
+ * sequence itself.
+ * <br>
+ * The information about whether a stored sequence is direct or not is not stored in the header of the Holder. Instead, the
+ * top bit of the pointer returned by {@link #ptr()} is used to store this information. This is done to save space in the
+ * header and to avoid the need to store this information separately. Thus, the value returned by {@link #ptr()} cannot
+ * be used as a pointer directly and should be treated as an opaque value.
+ * <p>
+ * Uses provided {@link GroupByAllocatorImpl} to allocate the underlying buffer. Grows the buffer when needed.
+ * <p>
+ * Buffer layout is the following:
+ * <pre>
+ * | capacity (in chars) | length (in chars) | char array or pointer/
+ * +---------------------+-------------------+----------------------+
+ * |       4 bytes       |      4 bytes      |           -          |
+ * +---------------------+-------------------+----------------------+
+ * </pre>
+ */
+public class DirectStrAwareCharHolder implements CharSequence {
+    private static final long HEADER_SIZE = 2 * Integer.BYTES;
+    private static final long LEN_OFFSET = Integer.BYTES;
+    private static final int MIN_CAPACITY = 4; // 4 chars = 8 bytes = enough to store a pointer
+    private GroupByAllocator allocator;
+    private long ptr;
+    private boolean direct;
+
+    /**
+     * Returns capacity in chars.
+     */
+    public int capacity() {
+        return ptr != 0 ? Unsafe.getUnsafe().getInt(ptr) : 0;
+    }
+
+    @Override
+    public char charAt(int index) {
+        if (direct) {
+            // we could cache the direct pointer, but then we would need to invalidate it when the pointer changes
+            // and we assume of() is called more frequently than charAt()
+            long directPtr = Unsafe.getUnsafe().getLong(ptr + HEADER_SIZE);
+            assert directPtr != 0;
+            assert index < length();
+            return Unsafe.getUnsafe().getChar(directPtr + 2L * index);
+        } else {
+            return Unsafe.getUnsafe().getChar(ptr + HEADER_SIZE + 2L * index);
+        }
+    }
+
+    private void clear() {
+        if (ptr != 0) {
+            Unsafe.getUnsafe().putInt(ptr + LEN_OFFSET, 0);
+            direct = false;
+        }
+    }
+
+    @Override
+    public int length() {
+        return ptr != 0 ? Unsafe.getUnsafe().getInt(ptr + LEN_OFFSET) : 0;
+    }
+
+    public DirectStrAwareCharHolder of(long ptr) {
+        // clear the top bit
+        this.ptr = ptr & 0x7FFFFFFFFFFFFFFFL;
+        // extract the top bit
+        this.direct = (ptr & 0x8000000000000000L) != 0;
+        return this;
+    }
+
+    public long ptr() {
+        return ptr | (direct ? 0x8000000000000000L : 0);
+    }
+
+    public void clearAndSet(@Nullable CharSequence cs) {
+        clear();
+        if (cs instanceof DirectCharSequence) {
```

Comment:
Probably, we need to check for a more specific interface/class here. That's because pointers are "stable" only for mmapped columns. If one of the SQL functions uses a char sink that implements `DirectCharSequence`, we may end up with segfaults first/last functions.

Reply:
that's a great point! any recommendation/idea what to check?

Reply:
it looks we would have to introduce a new marker interface?

Reply:
@puzpuzpuz I just pushed a change. let me know what you think.